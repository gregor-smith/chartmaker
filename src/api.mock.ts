import type { searchLastFM as _search } from './api.js'
export type { SearchResult, SearchArguments } from './api.js'


export const search = jest.fn<ReturnType<typeof _search>, Parameters<typeof _search>>()
