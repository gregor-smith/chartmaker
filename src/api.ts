import {
    Record as Record_,
    Array as Array_,
    String as String_,
    Static
} from 'runtypes'


const LastFMAlbum = Record_({
    name: String_,
    artist: String_,
    image: Array_(
        Record_({
            '#text': String_
        })
    )
})


const LastFMResult = Record_({
    results: Record_({
        albummatches: Record_({
            album: Array_(LastFMAlbum)
        })
    })
})


export type LastFMAlbum = Static<typeof LastFMAlbum>
export type LastFMResult = Static<typeof LastFMResult>


function isLastFMNullString(value: string): boolean {
    // Although Last.fm offers a JSON API, it doesn't make use of any JSON
    // types other than strings. Null values are represented by the string
    // '(null)' or just an empty string. Why even bother with JSON then...?
    return value === '' || value === '(null)'
}


export type SearchResultAlbum = {
    name: string
    url: string
}


function formatLastFMResult(result: LastFMResult): SearchResultAlbum[] {
    const albums: SearchResultAlbum[] = []
    for (const album of result.results.albummatches.album) {
        if (isLastFMNullString(album.artist)
                || isLastFMNullString(album.name)
                || album.image.some(image => isLastFMNullString(image['#text']))) {
            continue
        }
        albums.push({
            name: `${album.artist} - ${album.name}`,
            url: album.image[album.image.length - 1]['#text']
        })
    }
    return albums
}


export type SearchArguments = {
    key: string
    query: string
    signal: AbortSignal
}


export type SearchResult =
    | { tag: 'Ok', albums: SearchResultAlbum[] }
    | { tag: 'StatusError', status: number }
    | { tag: 'JSONDecodeError' }
    | { tag: 'InvalidResponseData' }
    | { tag: 'NetworkError' }
    | { tag: 'Cancelled' }


function joinURLQuery(base: string, query: Record<string, string>): string {
    const joinedQuery = Object.entries(query)
        .map(([ key, value ]) => {
            const escapedKey = encodeURIComponent(key.trim())
            const escapedValue = encodeURIComponent(value.trim())
            return `${escapedKey}=${escapedValue}`
        })
        .join('&')
    return `${base}?${joinedQuery}`
}


export async function search({
    key,
    query,
    signal
}: SearchArguments): Promise<SearchResult> {
    const url = joinURLQuery('https://ws.audioscrobbler.com/2.0/', {
        method: 'album.search',
        format: 'json',
        api_key: key,
        album: query
    })

    let response: Response
    try {
        response = await fetch(url, { signal })
    }
    catch {
        return {
            tag: signal.aborted
                ? 'Cancelled'
                : 'NetworkError'
        }
    }
    if (!response.ok) {
        return {
            tag: 'StatusError',
            status: response.status
        }
    }

    let result: unknown
    try {
        result = await response.json()
    }
    catch {
        return { tag: 'JSONDecodeError' }
    }

    if (!LastFMResult.guard(result)) {
        return { tag: 'InvalidResponseData' }
    }

    return {
        tag: 'Ok',
        albums: formatLastFMResult(result)
    }
}
