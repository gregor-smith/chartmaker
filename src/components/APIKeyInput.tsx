import type { FC } from 'react'

import type { DispatchProps } from '@/reducer'
import { SidebarGroup } from '@/components/SidebarGroup'
import { Label } from '@/components/Label'
import { ControlledInput } from '@/components/ControlledInput'


export const id = 'apiKeyInput'


export type APIKeyInputProps = DispatchProps & {
    apiKey: string
}


export const APIKeyInput: FC<APIKeyInputProps> = ({ dispatch, apiKey }) => {
    function updateAPIKey(apiKey: string) {
        dispatch({
            tag: 'UpdateAPIKey',
            apiKey
        })
    }

    return (
        <SidebarGroup>
            <Label target={id}>
                Last.fm API key
            </Label>
            <ControlledInput id={id}
                type='password'
                value={apiKey}
                onChange={updateAPIKey}/>
        </SidebarGroup>
    )
}
