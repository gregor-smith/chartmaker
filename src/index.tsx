import 'bootstrap/scss/bootstrap-reboot.scss'

import React from 'react'
import ReactDOM from 'react-dom'

import { App } from './App'


if (process.env.NODE_ENV === 'development') {
    localStorage.clear()
}


ReactDOM.render(
    <App/>,
    document.getElementById('react-root')
)
