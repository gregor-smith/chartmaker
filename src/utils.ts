import html2canvas from 'html2canvas'

import { BACKGROUND_COLOUR } from './style'


export function readFileText(file: File): Promise<string> {
    return new Promise((resolve, reject) => {
        const reader = new FileReader()
        reader.addEventListener('error', reject)
        reader.addEventListener('load', () => resolve(reader.result as string))
        reader.readAsText(file, 'utf-8')
    })
}


export function findIndex<T>(array: ReadonlyArray<T>, predicate: (item: T) => boolean): number | null {
    const index = array.findIndex(predicate)
    return index === -1 ? null : index
}


export async function elementToDataURI(element: HTMLElement, scale: number) {
    const canvas = await html2canvas(element, {
        allowTaint: true,
        scale,
        useCORS: true,
        imageTimeout: 0,
        backgroundColor: BACKGROUND_COLOUR
    })
    return canvas.toDataURL()
}


export function jsonToDataURI(json: string): string {
    return 'data:application/json;charset=utf-8,' + encodeURIComponent(json)
}


export function downloadURI(uri: string, filename: string) {
    const link = document.createElement('a')
    link.href = uri
    link.download = filename
    link.click()
    link.remove()
}
