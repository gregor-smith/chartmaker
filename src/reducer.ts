import { SideEffectUpdate, Dispatch as _Dispatch } from './hooks'
import {
    State,
    createChart,
    SearchState,
    Album,
    escapeStateForExport,
    ChartShape,
    NamedAlbum
} from './state'
import {
    readFileText,
    findIndex,
    elementToDataURI,
    downloadURI,
    jsonToDataURI
} from './utils'
import { search } from './api'


type Action =
    | { tag: 'UpdateAPIKey', apiKey: string }
    | { tag: 'UpdateActiveChart', index: number }
    | { tag: 'PromptForNewChart' }
    | { tag: 'ShowChartNameTakenMessage' }
    | { tag: 'AddNewChart', name: string }
    | { tag: 'PromptToRenameActiveChart' }
    | { tag: 'RenameActiveChart', name: string }
    | { tag: 'PromptToDeleteActiveChart' }
    | { tag: 'DeleteActiveChart' }
    | { tag: 'ImportStateFile', file: File }
    | { tag: 'ShowInvalidStateImportMessage' }
    | { tag: 'LoadState', state: State }
    | { tag: 'PromptToExportState' }
    | { tag: 'CancelSearchRequest' }
    | { tag: 'SendSearchRequest' }
    | { tag: 'UpdateSearchState', state: SearchState }
    | { tag: 'UpdateSearchQuery', query: string }
    | { tag: 'DragChartAlbum', sourceID: number, targetID: number }
    | { tag: 'DropSearchAlbum', sourceID: number, targetID: number }
    | { tag: 'PromptToRenameAlbum', id: number }
    | { tag: 'RenameAlbum', id: number, name: string }
    | { tag: 'DeleteAlbum', id: number }
    | { tag: 'UpdateScreenshotLoading', loading: boolean }
    | { tag: 'UpdateScreenshotScale', scale: number }
    | { tag: 'TakeScreenshot', element: HTMLElement }
    | { tag: 'UpdateChartShape', shape: ChartShape, rowsX: number, rowsY: number }


export type ActionTag = Action['tag']
export type Dispatch<T extends ActionTag = ActionTag> = _Dispatch<Extract<Action, { tag: T }>>
export type DispatchProps<T extends ActionTag = ActionTag> = {
    dispatch: Dispatch<T>
}


export function reducer(state: State, action: Action): SideEffectUpdate<State, Action> {
    switch (action.tag) {
        case 'UpdateAPIKey':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    apiKey: action.apiKey
                }
            }
        case 'UpdateActiveChart':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    activeChartIndex: action.index
                }
            }
        case 'PromptForNewChart':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const name = prompt('Enter new chart name:')?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    if (state.charts.some(chart => chart.name === name)) {
                        dispatch({ tag: 'ShowChartNameTakenMessage' })
                        return
                    }
                    dispatch({ tag: 'AddNewChart', name })
                }
            }
        case 'ShowChartNameTakenMessage':
            return {
                tag: 'SideEffect',
                sideEffect: () =>
                    alert('A chart with that name already exists')
            }
        case 'AddNewChart': {
            const [ albumIDCounter, chart ] = createChart({
                albumIDCounter: state.albumIDCounter,
                name: action.name
            })
            const charts = [ ...state.charts, chart ]
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts,
                    activeChartIndex: charts.length - 1,
                    albumIDCounter
                }
            }
        }
        case 'PromptToRenameActiveChart':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const activeChart = state.charts[state.activeChartIndex]
                    const name = prompt(`Enter new name for '${activeChart.name}':`)?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    if (state.charts.some((chart, index) => index !== state.activeChartIndex && chart.name === name)) {
                        dispatch({ tag: 'ShowChartNameTakenMessage' })
                        return
                    }
                    dispatch({ tag: 'RenameActiveChart', name })
                }
            }
        case 'RenameActiveChart': {
            const activeChart = state.charts[state.activeChartIndex]
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            name: action.name
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'PromptToDeleteActiveChart':
            return {
                tag: 'SideEffect',
                sideEffect: dispatch => {
                    if (confirm('Really delete active chart? This cannot be undone')) {
                        dispatch({ tag: 'DeleteActiveChart' })
                    }
                }
            }
        case 'DeleteActiveChart': {
            if (state.charts.length === 1) {
                const [ albumIDCounter, chart ] = createChart({
                    albumIDCounter: state.albumIDCounter
                })
                return {
                    tag: 'Update',
                    state: {
                        ...state,
                        charts: [ chart ],
                        activeChartIndex: 0,
                        albumIDCounter
                    }
                }
            }

            const charts = [
                ...state.charts.slice(0, state.activeChartIndex),
                ...state.charts.slice(state.activeChartIndex + 1)
            ]

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts,
                    activeChartIndex: state.activeChartIndex - 1 < 0
                        ? charts.length - 1
                        : state.activeChartIndex  -1
                }
            }
        }
        case 'ImportStateFile':
            return {
                tag: 'SideEffect',
                sideEffect: async dispatch => {
                    try {
                        const json = await readFileText(action.file)
                        const parsed: unknown = JSON.parse(json)
                        const state = State.check(parsed)
                        dispatch({ tag: 'LoadState', state })
                    }
                    catch {
                        dispatch({ tag: 'ShowInvalidStateImportMessage' })
                    }
                }
            }
        case 'ShowInvalidStateImportMessage':
            return {
                tag: 'SideEffect',
                sideEffect: () =>
                    alert('Selected file is invalid')
            }
        case 'LoadState':
            return {
                tag: 'Update',
                state: action.state
            }
        case 'PromptToExportState':
            return {
                tag: 'SideEffect',
                sideEffect: (_dispatch, state) => {
                    const json = JSON.stringify(escapeStateForExport(state))
                    const uri = jsonToDataURI(json)
                    downloadURI(uri, 'state.json')
                }
            }
        case 'CancelSearchRequest': {
            if (state.search.tag !== 'Loading') {
                return { tag: 'NoUpdate' }
            }
            const controller = state.search.controller
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    search: {
                        tag: 'Waiting',
                        query: state.search.query
                    }
                },
                sideEffect: () => controller.abort()
            }
        }
        case 'SendSearchRequest': {
            if (state.search.tag === 'Loading'
                    || state.search.query.trim().length === 0) {
                return { tag: 'NoUpdate' }
            }

            if (state.apiKey.trim().length === 0) {
                return {
                    tag: 'Update',
                    state: {
                        ...state,
                        search: {
                            tag: 'Error',
                            query: state.search.query,
                            message: 'Last.fm API key required'
                        }
                    }
                }
            }

            const controller = new AbortController()
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    search: {
                        tag: 'Loading',
                        query: state.search.query,
                        controller
                    }
                },
                sideEffect: async (dispatch, state) => {
                    const result = await search({
                        key: state.apiKey,
                        query: state.search.query,
                        signal: controller.signal
                    })
                    switch (result.tag) {
                        case 'Ok': {
                            const albums: NamedAlbum[] = result.albums.map((album, index) => ({
                                ...album,
                                placeholder: false,
                                id: state.albumIDCounter + index + 1
                            }))
                            dispatch({
                                tag: 'LoadState',
                                state: {
                                    ...state,
                                    search: {
                                        tag: 'Complete',
                                        query: state.search.query,
                                        albums
                                    },
                                    albumIDCounter: state.albumIDCounter + result.albums.length
                                }
                            })
                            break
                        }
                        case 'StatusError': {
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: `Last.fm request returned ${result.status}`
                                }
                            })
                            break
                        }
                        case 'JSONDecodeError': {
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: 'Invalid data returned from Last.fm'
                                }
                            })
                            break
                        }
                        case 'NetworkError':
                            dispatch({
                                tag: 'UpdateSearchState',
                                state: {
                                    tag: 'Error',
                                    query: state.search.query,
                                    message: 'Network error sending request to Last.fm'
                                }
                            })
                        // 'Cancelled' is not handled because the only time the
                        // request is cancelled is when another is sent, in
                        // which case the search state is updated to 'Loading'
                        // anyway
                    }
                }
            }
        }
        case 'UpdateSearchState':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    search: action.state
                }
            }
        case 'UpdateSearchQuery': {
            if (state.search.tag === 'Loading') {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'Update',
                state: {
                    ...state,
                    search: {
                        ...state.search,
                        query: action.query
                    }
                }
            }
        }
        case 'DragChartAlbum': {
            if (action.sourceID === action.targetID) {
                return { tag: 'NoUpdate' }
            }

            const activeChart = state.charts[state.activeChartIndex]

            const sourceIndex = findIndex(activeChart.albums, album => album.id === action.sourceID)
            if (sourceIndex === null) {
                return { tag: 'NoUpdate' }
            }

            const targetIndex = findIndex(activeChart.albums, album => album.id === action.targetID)
            if (targetIndex === null) {
                return { tag: 'NoUpdate' }
            }

            let albums: Album[]
            if (sourceIndex < targetIndex) {
                albums = [
                    ...activeChart.albums.slice(0, sourceIndex),
                    ...activeChart.albums.slice(sourceIndex + 1, targetIndex + 1),
                    activeChart.albums[sourceIndex],
                    ...activeChart.albums.slice(targetIndex + 1)
                ]
            }
            else {
                albums = [
                    ...activeChart.albums.slice(0, targetIndex),
                    activeChart.albums[sourceIndex],
                    ...activeChart.albums.slice(targetIndex, sourceIndex),
                    ...activeChart.albums.slice(sourceIndex + 1)
                ]
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'DropSearchAlbum': {
            if (state.search.tag !== 'Complete') {
                return { tag: 'NoUpdate' }
            }

            const activeChart = state.charts[state.activeChartIndex]

            const source = state.search.albums.find(album => album.id === action.sourceID)
            if (source === undefined) {
                return { tag: 'NoUpdate' }
            }

            const targetIndex = findIndex(activeChart.albums, album => album.id === action.targetID)
            if (targetIndex === null) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, targetIndex),
                                {
                                    ...source,
                                    id: state.albumIDCounter + 1
                                },
                                ...activeChart.albums.slice(targetIndex + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ],
                    albumIDCounter: state.albumIDCounter + 1
                }
            }
        }
        case 'PromptToRenameAlbum':
            return {
                tag: 'SideEffect',
                sideEffect: (dispatch, state) => {
                    const album = state.charts[state.activeChartIndex]
                        .albums
                        .find(album => album.id === action.id)
                    if (album === undefined || album.placeholder) {
                        return
                    }

                    const name = prompt(`Enter new name for '${album.name}':`)?.trim()
                    if (name === undefined || name.length === 0) {
                        return
                    }
                    dispatch({
                        tag: 'RenameAlbum',
                        id: action.id,
                        name
                    })
                }
            }
        case 'RenameAlbum': {
            const activeChart = state.charts[state.activeChartIndex]

            const index = findIndex(activeChart.albums, album => album.id === action.id)
            if (index === null) {
                return { tag: 'NoUpdate' }
            }

            const album = activeChart.albums[index]
            if (album.placeholder) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, index),
                                {
                                    ...album,
                                    name: action.name
                                },
                                ...activeChart.albums.slice(index + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'DeleteAlbum': {
            const activeChart = state.charts[state.activeChartIndex]

            const index = findIndex(activeChart.albums, album => album.id === action.id)
            if (index === null) {
                return { tag: 'NoUpdate' }
            }

            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...activeChart,
                            albums: [
                                ...activeChart.albums.slice(0, index),
                                {
                                    id: action.id,
                                    placeholder: true
                                },
                                ...activeChart.albums.slice(index + 1)
                            ]
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
        }
        case 'UpdateScreenshotLoading': {
            return {
                tag: 'Update',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        loading: action.loading
                    }
                }
            }
        }
        case 'UpdateScreenshotScale': {
            if (state.screenshot.loading) {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'Update',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        scale: action.scale
                    }
                }
            }
        }
        case 'TakeScreenshot': {
            if (state.screenshot.loading) {
                return { tag: 'NoUpdate' }
            }
            return {
                tag: 'UpdateWithSideEffect',
                state: {
                    ...state,
                    screenshot: {
                        ...state.screenshot,
                        loading: true
                    }
                },
                sideEffect: async (dispatch, state) => {
                    const uri = await elementToDataURI(
                        action.element,
                        state.screenshot.scale
                    )
                    downloadURI(uri, 'chart.png')
                    dispatch({
                        tag: 'UpdateScreenshotLoading',
                        loading: false
                    })
                }
            }
        }
        case 'UpdateChartShape':
            return {
                tag: 'Update',
                state: {
                    ...state,
                    charts: [
                        ...state.charts.slice(0, state.activeChartIndex),
                        {
                            ...state.charts[state.activeChartIndex],
                            shape: action.shape,
                            rowsX: action.rowsX,
                            rowsY: action.rowsY
                        },
                        ...state.charts.slice(state.activeChartIndex + 1)
                    ]
                }
            }
    }
}
