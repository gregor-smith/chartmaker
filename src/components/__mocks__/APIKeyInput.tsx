import React, { FC } from 'react'

import { APIKeyInputProps } from '@/components/APIKeyInput'


export const APIKeyInput: FC<APIKeyInputProps> = ({ apiKey }) =>
    <div className='mock-api-key-input'>
        {`API Key: ${apiKey}`}
    </div>
