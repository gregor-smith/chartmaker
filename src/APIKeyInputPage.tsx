import React, { FC, ChangeEvent, FormEvent } from 'react'
import { css } from 'emotion'

import { DispatchProps } from './reducer'


export type APIKeyInputPageState = {
    key: string
}


type Props = DispatchProps<'UpdateAPIKeyInputText' | 'ChangeAPIKey'> & {
    text: string
}


const pageStyle = css({
    display: 'flex',
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    flexGrow: 1
})


const formStyle = css({
    display: 'flex',
    padding: '1rem',
    border: '1px solid white'
})


export const APIKeyInputPage: FC<Props> = ({ dispatch, text: key }) => {
    function updateKey(event: ChangeEvent<HTMLInputElement>) {
        event.preventDefault()
        dispatch({
            tag: 'UpdateAPIKeyInputText',
            key: event.currentTarget.value
        })
    }

    function confirmKey(event: FormEvent<HTMLFormElement>) {
        event.preventDefault()
        dispatch({ tag: 'ChangeAPIKey', key })
    }

    return (
        <main className={pageStyle}>
            <form className={formStyle} onSubmit={confirmKey}>
                <input type='password'
                    placeholder='Enter Last.fm API key'
                    autoFocus
                    onChange={updateKey}/>
                <button type='submit'>
                    Confirm
                </button>
            </form>
        </main>
    )
}
