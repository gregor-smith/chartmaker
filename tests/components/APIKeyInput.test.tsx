import React, { createRef, MutableRefObject } from 'react'
import { render } from 'react-dom'
import { act } from 'react-dom/test-utils'

import { setRenderContainer, clearRenderContainer, ignore, fireEvent } from '../utils'
import APIKeyInput from '../../src/components/APIKeyInput'
import { Action } from '../../src/reducer'


const container: MutableRefObject<HTMLElement | null> = createRef()
beforeEach(() => setRenderContainer(container))
afterEach(() => clearRenderContainer(container))


// this line breaks vscode's highlights despite being syntactically correct. gg
type UpdateAPIKeyAction = Extract<Action, { tag: 'UpdateAPIKey' }>


test('renders labelled input', () => {
    render(
        <APIKeyInput dispatch={ignore} apiKey='test api key'/>,
        container.current
    )
    expect(container.current).toMatchSnapshot()
})


test('dispatches action on input change', () => {
    const mock = jest.fn<void, [ UpdateAPIKeyAction ]>()
    render(
        <APIKeyInput dispatch={mock} apiKey='test api key'/>,
        container.current
    )

    act(() => {
        const input = container.current?.querySelector('input')
        if (input == null) {
            return
        }
        input.value = 'test new api key'
        fireEvent('change', input)
    })

    expect(mock).toHaveBeenCalledTimes(1)
    expect(mock).toHaveBeenCalledWith<[ UpdateAPIKeyAction ]>({
        tag: 'UpdateAPIKey',
        apiKey: 'test new api key'
    })
})
