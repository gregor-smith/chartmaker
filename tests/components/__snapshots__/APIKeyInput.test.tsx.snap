// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders labelled input 1`] = `
.emotion-1 {
  background: black;
  color: white;
  border: 1px solid white;
  padding: 0.25rem;
  width: 100%;
}

.emotion-2:not(:last-of-type) {
  border-bottom: 1px solid white;
  padding-bottom: 1rem;
  margin-bottom: 1rem;
}

.emotion-0 {
  margin-bottom: 0.25rem;
}

<div
  id="test-render-container"
>
  <div
    class="emotion-2"
  >
    <label
      class="emotion-0"
      for="apiKeyInput"
    >
      Last.fm API key
    </label>
    <input
      class="emotion-1"
      id="apiKeyInput"
      type="password"
      value="test api key"
    />
  </div>
</div>
`;
